### PR TITLE
Multi file

### PR DESCRIPTION
**Pull request by hildjj**
This is a PR draft exploring another part of #292.  

What if you could just pass multiple files in on the command line, and have Peggy stitch them together into a single output?

This points out a few issues:

- Code generation does not rely on source information in the AST adequately.  Instead, it guesses based on the grammarSource passed in.  This is another strong argument for changing how we do code generation.
- `topLevelInitializer` and `initializer` in the AST should be arrays.  This is so that when they get copied into the output, each initializer section would be marked up with the correct source-map info.
- There might be `import`s in the top-level initializers of multiple files.  They should probably all be moved to the top, but this is a JS-specific transform that should be in a compiler pass.

### CODE REVIEW

**Comment by hildjj** [comment]
There's a timing bug in the watchfile tests as well.  Will deal with it if we end up using this.

**Comment by hildjj** [comment]
I think the sourcemap thing isn't an issue.  This works surprisingly well already.  Still needs a compiler pass that checks if output type is "es" and looks for import lines to move to the top.

**Comment by hildjj** [comment]
Still needs docs.

**Comment by hildjj** [comment]
Being able to specify a URL in the CLI would be pretty cool.  In order to do this, I'd look for filenames starting with `https?://`, use fetch to pull down the contents, and cache it, like deno.  This would up the minimum version of node required to be v18 -- That's fine with me because:

a) we're taking a breaking change in this release for node version anyway.
b) Node 16 is [no longer supported](https://nodejs.org/en/blog/announcements/nodejs16-eol).

This would make importing rules from a central repository easier, I think.

File: bin/peggy-cli.js [hunk 1]
```diff
@@ -115,7 +115,7 @@ class PeggyCLI extends Command {
 
     this
       .version(peggy.VERSION, "-v, --version")
-      .argument("[input_file...]", 'Grammar file(s) to read.  Use "-" to read stdin.  If multiple files are given, they are combined in the given order to produce a single output.', ["-"])
+      .argument("[input_file...]", 'Grammar file(s) to read.  Use "-" to read stdin.  If multiple files are given, they are combined in the given order to produce a single output.  Use npm:"<pacakgeName>/file.peggy" to import from an npm dependency.', ["-"])
```
Comment (Mingun) [inline bin/peggy-cli.js:118]: Misprint in `packageName`:
```suggestion
      .argument("[input_file...]", 'Grammar file(s) to read.  Use "-" to read stdin.  If multiple files are given, they are combined in the given order to produce a single output.  Use npm:"<packageName>/file.peggy" to import from an npm dependency.', ["-"])
```

Comment (hildjj) [reply bin/peggy-cli.js]: Good catch.  Fixed, along with the matching test.

**Comment by frostburn** [comment]
Tested combining grammars from two repositories successfully:  https://github.com/frostburn/peggy-string-concat

File: bin/peggy-cli.js [hunk 2]
```diff
@@ -284,21 +284,31 @@ class PeggyCLI extends Command {
...
-          this.inputFile = this.progOptions.input;
+          this.inputFiles = Array.isArray(this.progOptions.input)
+            ? this.progOptions.input
+            : [this.progOptions.input];
         }
         this.outputFile = this.progOptions.output;
         this.outputJS = this.progOptions.output;
 
-        if ((this.inputFile === "-") && this.argv.watch) {
+        if ((this.inputFiles.indexOf("-") !== -1) && this.argv.watch) {
```
Comment (frostburn) [inline bin/peggy-cli.js:294]: How about the more readable `inputFiles.includes("-")` insteads?

File: bin/peggy-cli.js [hunk 3]
```diff
@@ -629,26 +639,40 @@ class PeggyCLI extends Command {
...
+        this.verbose("CLI", errorText = `reading input "${source}"`);
+        if (source === "-") {
+          input.source = "stdin";
+          this.std.in.resume();
+          input.text = await readStream(this.std.in);
+        } else if (source.startsWith("npm:")) {
+          const req = (
+            // In node 12+, createRequire is documented.
+            // In node 10, createRequireFromPath is the least-undocumented approach.
+            Module.createRequire || Module.createRequireFromPath
```
Comment (frostburn) [inline bin/peggy-cli.js:659]: Using nullish coalescing `??` would be more idiomatic here. Are we on a node version that doesn't support it? Not seeing it anywhere in the source.

**Review by Mingun — COMMENTED** [review]
The implementation looks good, but I don't like the fact that peggy grammar becomes tied to JS. That is the same problem that we have now for unbalanced `{}` which could be present in JS comments or strings and actually does not create unbalanced braces, but peggy grammar does not understand that because code blocks are language-agnostic.

I suggest to merge this without changes in the grammar and propose grammar changes in separate PR. Ideally we should find a way to make grammar modularised so language plugins can provide they own methods to parse code blocks.

File: bin/peggy-cli.js (hunk 3 above)
Comment (hildjj) [reply bin/peggy-cli.js:659]: I think createRequire is in node12, but I'm fine bumping to Node18 required, and using anything new we want in the CLI code.  The web-facing code we should be a little more careful with.

File: src/parser.pegjs [hunk 4]
```diff
@@ -57,14 +57,64 @@ Grammar
     }
 
 TopLevelInitializer
-  = "{" code:CodeBlock "}" EOS {
-      return {
+  = "{" "{" imports:ImportDeclarations code:BareCodeBlock "}" "}" EOS {
```
[the remark this answers is not in the document]
Comment (hildjj) [reply src/parser.pegjs:60]: Yes, I agree this is an issue.  I suppose we could have a sub-grammar that can be used in generate-js to perform this transform.

**Comment by hildjj** [comment]
Another patch coming with other issues fixed.

**Comment by hildjj** [comment]
OK, I think I fixed all of the outstanding issues.  I've left this as two patches since review to make it easier to see the changes, but I'll reorder and clean up the patch before merging.

File: src/parser.pegjs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hildjj) [reply src/parser.pegjs]: I've implemented the sub-grammar approach, which only gets called from generate-js, and only if we are in "es" format.  This should be a lot safer.

File: lib/compiler/asts.js [hunk 5]
```diff
@@ -2,8 +2,34 @@
...
+ * @returns {T | T[]}
+ */
+function combinePossibleArrays(a, b) {
+  const aa = Array.isArray(a) ? a : [a];
+  const bb = Array.isArray(b) ? b : [b];
+
+  // Put library code above code that uses it, so that variables
+  // will be in scope.
+  const res = bb.filter(x => x).concat(aa.filter(x => x)); // Remove nulls.
+  return res;
```
[the remark this answers is not in the document]
Comment (hildjj) [reply lib/compiler/asts.js]: Agree.  As well, if we aren't generating es-js, the inputs should not be reordered.  That's a decision for the generator.

File: docs/documentation.html [hunk 6]
```diff
@@ -126,6 +126,13 @@ <h3 id="generating-a-parser-command-line">Command Line</h3>
 <p>If you omit both input and output file, standard input and standard output
   are used.</p>
 
+<p>If you specify multiple input files, they will be folded together in the
+  order specified before generating a parser.  <code>import</code> statements in
+  the top-level initializers from each of the inputs will be moved to the top of
+  the generated code, and all other top-level initializers will be inserted
+  directly after those imports, in the order of the inputs.  This approach can
+  be used to keep libraries of often-used grammar rules in separate files.</p>
+
```
[the remark this answers is not in the document]
Comment (hildjj) [reply docs/documentation.html]: See if the comment I added was good enough?  Importantly, the ECMAscript grammar treats all comments after an import as part of the import.

File: bin/peggy-cli.js (hunk 3 above)
Comment (hildjj) [reply bin/peggy-cli.js]: Just removed node 10 compatibility.

File: src/js-imports.peggy [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ModuleSpecifier = StringLiteral
+ImportedBinding
+  = BindingIdentifier
+
+ModuleExportName
+  =  IdentifierName
+  / StringLiteral
+
+BindingIdentifier = id:IdentifierName &{
+  return reservedWords.indexOf(id[0]) === -1
```
Comment (frostburn) [inline src/js-imports.peggy:47]: `.includes` instead

File: bin/peggy-cli.js [hunk 8]
```diff
@@ -284,21 +284,31 @@ class PeggyCLI extends Command {
...
-        if ((this.inputFile === "-") && this.argv.watch) {
+        if ((this.inputFiles.includes("-")) && this.argv.watch) {
           this.argv.watch = false; // Make error throw.
           this.error("Can't watch stdin");
         }
 
         if (!this.outputFile) {
-          if (this.inputFile !== "-") {
-            this.outputJS = this.inputFile.slice(
+          if (this.inputFiles.indexOf("-") === -1) {
```
Comment (frostburn) [inline bin/peggy-cli.js:300]: `.includes` instead

File: src/js-imports.peggy (hunk 7 above)
Comment (hildjj) [reply src/js-imports.peggy:47]: No, this one needs to run in the browser.

**Comment by hildjj** [comment]
This is ready to squash and merge.

**Comment by hildjj** [comment]
ready to go if anyone wants to take a last look.  Next step is to merge the imports grammar back into the main grammar, and give it two entry points.  Then, I'll use the imports grammar to add imports syntax before toplevel initializers.

**Pull request merged**